File: src/main/java/com/simprints/libsimprints/Identification.java
package com.simprints.libsimprints;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

@SuppressWarnings("unused")
public class Identification implements Parcelable, Comparable<Identification> {

    private String guid;
    private int confidence;
    private Tier tier;

    /**
     * Empty constructor
     */
    public Identification() {

    }

    /**
     * This constructor creates a new identification
     *
     * @param guid       A string containing the guid
     * @param confidence An int containing the (matching) confidence
     */
    public Identification(String guid, int confidence, Tier tier) {
        this.guid = guid;
        this.confidence = confidence;
        this.tier = tier;
    }

    protected Identification(Parcel in) {
        guid = in.readString();
        confidence = in.readInt();
        tier = Tier.values()[in.readInt()];
    }

    public static final Creator<Identification> CREATOR = new Creator<Identification>() {
        @Override
        public Identification createFromParcel(Parcel in) {
            return new Identification(in);
        }

        @Override
        public Identification[] newArray(int size) {
            return new Identification[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.guid);
        dest.writeInt(this.confidence);
        dest.writeInt(this.tier.ordinal());
    }

    public String getGuid() {
        return guid;
    }

    public void setGuid(String guid) {
        this.guid = guid;
    }

    public float getConfidence() {
        return confidence;
    }

    public void setConfidence(int confidence) {
        this.confidence = confidence;
    }

    public Tier getTier() {
        return tier;
    }

    public void setTier(Tier tier) {
        this.tier = tier;
    }

    @Override
    public int compareTo(@NonNull Identification otherId) {
        if (confidence == otherId.confidence) {
            return 0;
        } else if (confidence < otherId.confidence) {
            return 1;
        } else {
            return -1;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (o instanceof Identification) {
            Identification other = (Identification)o;
            return (guid.equals(other.guid) &&
                    confidence == other.confidence &&
                    tier.equals(other.tier));
        }
        return false;
    }

    @Override
    public int hashCode() {
        int hash = guid.hashCode();
        hash = hash * 17 + confidence;
        hash = hash * 17 + tier.hashCode();
        return hash;
    }
}

File: src/main/java/com/simprints/libsimprints/Identification.kt
package com.simprints.libsimprints

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator

@Deprecated("Use contracts.data.Identification instead")
data class Identification(
    val guid: String,
    private val confidence: Int,
    val tier: Tier,
) : Parcelable,
    Comparable<Identification> {
    fun getConfidence(): Float = confidence.toFloat()

    override fun compareTo(other: Identification): Int = when {
        confidence == other.confidence -> 0
        confidence < other.confidence -> 1
        else -> -1
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(
        dest: Parcel,
        flags: Int,
    ) {
        dest.writeString(guid)
        dest.writeInt(this.confidence)
        dest.writeInt(this.tier.ordinal)
    }

    companion object {
        @JvmField
        val CREATOR = object : Creator<Identification> {
            override fun createFromParcel(parcel: Parcel): Identification? {
                val guid = parcel.readString().orEmpty()
                val confidence = parcel.readInt()
                val tier = Tier.entries[parcel.readInt()]

                return Identification(guid, confidence, tier)
            }

            override fun newArray(p0: Int): Array<out Identification?> = arrayOfNulls(p0)
        }
    }
}
